cfgov/v1/storage.py

```diff
@@ -0,0 +1,120 @@
+    maintain a consistent relationship between source filename and filename
+    in storage.
+
+    Wagtail 1.13 included changes that broke support for backends with this
+    overwriting behavior for both documents and images:
+
+    https://github.com/wagtail/wagtail/pull/3821
+    https://github.com/wagtail/wagtail/pull/3822
+
+    The goal of these changes was to clean up old document/image files after
+    the user uploads a new one. Consider an existing document uploaded from
+    some filename "test.txt"; the source file was "test.txt" and the file on
+    the storage backend will be "test.txt". If a user uses the Wagtail UI to
+    replace the document's file with a new file "test2.txt", the change in
+    1.13 tries to clean up the old file by deleting "test.txt" from the storage
+    backend after "test2.txt" has been stored.
+
+    The Wagtail logic looks roughly like this:
+
+        - user uploads a new file for an existing document/image
+        - store the new file to the storage backend
+        - delete the document/image's old file from the storage backend
+
+    This logic is sound as long as it is safe to assume that new uploads with
+    identical filenames never overwrite each other and always get assigned a
+    unique filename when stored. However, as described above, this may not
+    always be the case.
+
+    If, instead, files get overwritten, then the final delete that happens
+    will in fact delete the new file that's just been stored.
+
+    This mixin attempts to prevent this from happening. It uses the Python
+    stack inspect capability to identify if a delete call is being invoked
+    from the Wagtail document/image edit views, and, if so, skips that delete.
+    """
+    def delete(self, name):
+        stack = inspect.stack()
+
+        if self._should_skip_wagtail_delete(stack):
+            return
+
+        super(WagtailOverwritingStorageMixin, self).delete(name)
+
+    def _should_skip_wagtail_delete(self, stack):
+        """Given a call stack, determine whether delete should be skipped."""
+
+        # See Python inspect documentation for structure of stack object:
+        # https://docs.python.org/2/library/inspect.html#the-interpreter-stack
+        caller_frame_record = stack[1]
+        caller_frame_object = caller_frame_record[0]
+        caller_frame_function_name = caller_frame_record[3]
+
+        caller_module = inspect.getmodule(caller_frame_object)
+        caller_module_name = caller_module.__name__
+
+        called_from = caller_module_name + '.' + caller_frame_function_name
+
+        # Only skip delete if it is being called by the wagtaildocs.edit or
+        # wagtailimages.edit views. We want deletes called from other places
+        # to work as expected.
+        skip_for_these_callers = (
+            'wagtail.wagtaildocs.views.documents.edit',
+            'wagtail.wagtailimages.views.images.edit',
+        )
```

Comment:
Would it be possible to call out where those other places are, explicitly? If we're going to do stack manipulation to force behavior we want, I think I'd prefer to explicitly enable it for callers, rather than vice-versa.

Reply:
I'm not sure I follow -- maybe this variable is just named poorly, but this is the list of places where we want to skip delete -- otherwise the default behavior works. We need to maintain the default functionality of `Storage.delete` in any other cases where it is used.

Reply:
Then does this work?

If we want to skip delete for `wagtail.wagtaildocs.views.documents.edit`, then shouldn't [line 88](https://github.com/cfpb/cfgov-refresh/pull/4195/files#diff-255059ee1b4d67208bda88e26493190bR88) return `True` so that [line 60](https://github.com/cfpb/cfgov-refresh/pull/4195/files#diff-255059ee1b4d67208bda88e26493190bR60) triggers the return before the super call?

Reply:
More evidence that I named these things poorly.

```py
if called_from not in skip_for_these_callers
```

This means

> if `delete` was called from a caller that isn't in the list of callers to skip `delete` for

meaning, we know we shouldn't skip delete, so `_should_skip_wagtail_delete` should return falsey.